### DISCUSSION TITLE
Changing timestepper type during a solve.

### DISCUSSION THREAD

**Discussion by EdSheltonUKAEA** [idaholab/moose#24028] [Q&A Modules: General] [answered]
Is there a way to change timestepper type during a solve?  E.g. start with a FunctionDT for exactly defined time steps, and then allow an IterationAdaptive timestepper to take over at a defined point.

start_time and end_time in the Transient Executioner block apply to all settings, not just the timestepper.

Can this be done using TimePeriod controls?  If so, how are these cross-referenced between the Executioner and Controls blocks?
Or can sub-blocks with appropriate start and end times be specified within the main Executioners block (similar to defining sub-blocks for a stitched mesh)?

I can't seem to find any specific examples of the syntax showing how to do these if they are possible.

Alternatively, do the parameters "time_t" and "time_dt" allow some specific timestep values to be set within an IterationAdaptiveDT timestepper?  Is there an example of this syntax?

These two previous questions and answers seem to have related information, but are for more specific cases reacting to values obtained during a solve:
https://github.com/idaholab/moose/discussions/23344
https://github.com/idaholab/moose/discussions/23414

I'm trying to set up some defined times at the start of a solve, so this should be much simpler (I think).

**Comment by GiudGiud** [comment]
Hello

Not currently. You currently cannot build two timesteppers. 
However @MengnanLi91 is designing a composition time stepper that will bring a lot of flexibility to this, and will less you specify more than one time stepper. 
See https://github.com/idaholab/moose/pull/23961

Guillaume

**Reply by EdSheltonUKAEA** [reply]
OK.  Thanks for this.  This should be a useful addition.

In the meantime is there an example using the parameters "time_t" and "time_dt" in the existing IterationAdaptiveDT timestepper?  I'm not sure if these will help me do what I want.

**Reply by GiudGiud** [reply]
There's an example here
(modules/tensor_mechanics/test/tests/anisotropic_elastoplasticity/ad_uniaxial_x.i)
does it work as expected?

**Reply by EdSheltonUKAEA** [reply]
Unfortunately, this isn't doing what I'd hoped for.  I expected the "time_dt" values to be default time step values which are applied after each corresponding "time_t" value.  But this isn't what happened.

It might easier to explain what I'd like to happen.

Initially, up to t = 600, I'd like the equivalent of this FunctionDT:
>   type = FunctionDT
>   function = 'if(t<60, 6, 30)'

Then after this, from t = 600:
>     type = IterationAdaptiveDT
>     dt = 24
>     cutback_factor = 0.5
>     growth_factor = 2

So the timestep could be cut back to 12, 6 (or smaller), but if possible it will grow, maybe applying a maximum value of 192 (or larger depending on circumstances).

I tried to do a version of this using:
>   [TimeStepper]
>     type = IterationAdaptiveDT
>     dt = 6
>     cutback_factor = 0.5
>     growth_factor = 2
>     time_t = '0 60 600 624'
>     time_dt = '6 30 6 24'
>   [] 

And I got this set of time values up to 600.  So the time step is growing and shrinking by factors which aren't 2 or 0.5.

> time
> --
> 0
> 6
> 14.4
> 26.16
> 42.624
> 60
> 90
> 118.6667
> 146.0593
> 172.2344
> 197.2462
> 221.1464
> 243.9843
> 265.8072
> 286.6602
> 306.5865
> 325.6271
> 343.8214
> 361.2071
> 377.8201
> 393.6948
> 408.8639
> 423.3589
> 437.2096
> 450.4447
> 463.0916
> 475.1764
> 486.7241
> 497.7586
> 508.3027
> 518.3781
> 528.0058
> 537.2055
> 545.9964
> 554.3965
> 562.4234
> 570.0934
> 577.4226
> 584.426
> 591.1182
> 597.513
> 600

Above 600, it did move between, 3, 6, 12 and 24.  But wouldn't go above 24.  So I have misunderstood what the array of values specified in the parameters do.

Can what I want be achieved using:
> force_step_every_function_point
> timestep_limiting_function
> max_function_change

**Reply by GiudGiud** [reply]
Use this as the limiting function

 'if(t<60, 6, 30)'

then once you are done with limiting, just use another if to make it blow away

if(t<600, if(t<60, 6, 30),    10000000)

**Reply by EdSheltonUKAEA** [reply]
It says:
> *** ERROR ***
> The following error occurred in the object "TimeStepper", of type "IterationAdaptiveDT".
> 
> timestep_limiting_function must be a PiecewiseBase function

So it doesn't seem I can use a ParsedFunction with nested ifs.

My broader query was why the timestep increased and decreased by factors other than the specified 2 and 0.5, and then didn't increase to higher timestep values even though the model was solving OK.

**Reply by EdSheltonUKAEA** [reply, marked as answer]
I've found a way to do it.
I couldn't use "ifs" with a timestep_limiting_function, but I can with a timestep_limiting_postprocessor using a parsed function.

So this TimeStepper block and associated PPs do something like what I wanted.

>   [TimeStepper] # Sub-block to vary default timestep during different stages of solve.
>     type = IterationAdaptiveDT
>     optimal_iterations = 6
>     iteration_window = 3
>     linear_iteration_ratio = 3
>     growth_factor = 2
>     cutback_factor = 0.5
>     dt = 6
>     timestep_limiting_postprocessor = P999_tslim_pp
>   []

> [Postprocessors]
>   [P999_tslim_pp]
>     type = ParsedPostprocessor
>     function = 'if(t<60, 6, 
>       if(t<540, 60,
>         if(t<660, 6, 1536)))'
>     pp_names = 'P998_dummy_pp'
>     use_t = true
>   []
>   [P998_dummy_pp]
>     type = EmptyPostprocessor
>   []
> []

The timesteps are controlled up to 660s, and held at 6s for 600s +/- 60s where I'm changing something in my model.
Then after 660s the brakes are taken off and the full adaptive timestep capability is allowed to let rip.

**Reply by GiudGiud** [reply]
ah I m glad you found a way. I think we will be able to do this easier in the future with the CompositionDT. 

Best,
Guillaume
